src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs

```diff
@@ -153,6 +155,68 @@ public async Task ManyManyConcatenatedGzipStreams(int streamCount, TestScenario
             await TestConcatenatedGzipStreams(streamCount, scenario, bufferSize, bytesPerStream);
         }
 
+        [Theory]
+        [InlineData(TestScenario.ReadAsync)]
+        [InlineData(TestScenario.Read)]
+        [InlineData(TestScenario.Copy)]
+        [InlineData(TestScenario.CopyAsync)]
+        public async Task StreamTruncation_IsDetected(TestScenario scenario)
+        {
+            var source = Enumerable.Range(0, 32).Select(i => (byte)i).ToArray();
+            byte[] data;
+            using (var compressed = new MemoryStream())
+            using (var gzip = new GZipStream(compressed, CompressionLevel.Fastest))
+            {
+                foreach (var b in source)
+                {
+                    gzip.WriteByte(b);
+                }
+
+                gzip.Dispose();
+                data = compressed.ToArray();
+            }
+
+            for (var i = 1; i < data.Length; i++)
+            {
+                using (var compressedStream = new MemoryStream(data.Take(i).ToArray()))
+                {
+                    using (var s = new GZipStream(compressedStream, CompressionMode.Decompress))
+                    {
+                        var buffer = new byte[1024];
+                        var decompressedStream = new MemoryStream();
+
+                        try
+                        {
+                            switch (scenario)
+                            {
+                                case TestScenario.Copy:
+                                    s.CopyTo(decompressedStream);
+                                    break;
+
+                                case TestScenario.CopyAsync:
+                                    await s.CopyToAsync(decompressedStream);
+                                    break;
+
+                                case TestScenario.Read:
+                                    while (s.Read(buffer, 0, buffer.Length) != 0) { };
+                                    break;
+
+                                case TestScenario.ReadAsync:
+                                    while (await s.ReadAsync(buffer, 0, buffer.Length) != 0) { };
+                                    break;
+                            }
+                        }
+                        catch (InvalidDataException)
+                        {
+                            continue;
+                        }
+
+                        throw new XunitException($"Truncated stream was decompressed successfully: length={i}");
```

Comment:
Aside of truncation tests, could we test changing byte values?

Reply:
sure, I'll re-use the test from the [original SO thread](https://stackoverflow.com/q/9456563/4064749), if that is fine with you.

Reply:
@mfkl it's fine to reuse that from SO per the SO license. If you end up essentially pasting as-is, please add a few lines to [our THIRD-PARTY-NOTICES](https://github.com/dotnet/runtime/blob/main/THIRD-PARTY-NOTICES.TXT) to ensure that the poster is recognized. If you write something that effectively does the same thing, that's probably not necessary.